syntheseus/search/algorithms/best_first/base.py

```diff
@@ -54,72 +119,86 @@ def _run_from_graph_after_setup(self, graph: GraphType) -> int:
-            latest_priority = self.priority_function(node, graph)
-
-            # Decide between 3 options: discarding the node, re-inserting it,
-            # or visiting it (which most likely means expanding it)
-            if not self.node_eligible_for_queue(node, graph):
-                action = "discarded (already expanded/not eligible for queue)"
-            elif not math.isclose(priority, latest_priority):
-                # Re-insert the node with the correct priority,
-                # unless the new priority is inf
-                priority_change_str = f"(priority changed from {priority} to {latest_priority})"
-                if latest_priority < math.inf:
-                    action = f"re-inserted {priority_change_str}"
-                    heapq.heappush(queue, (latest_priority, tie_breaker, node))
-                    tie_breaker += 1
-                else:
-                    action = f"discarded {priority_change_str}"
-            else:
-                # Visit node
-                new_nodes = list(self.visit_node(node, graph))
-
-                # Update node values
-                nodes_updated = self.set_node_values(new_nodes + [node], graph)
-
-                # Add new eligible nodes to the queue, since
-                # their priority may have changed.
-                # dict.fromkeys is to preserve order and uniqueness
-                for updated_node in dict.fromkeys(new_nodes + list(nodes_updated)):
-                    if self.node_eligible_for_queue(updated_node, graph):
-                        updated_node_priority = self.priority_function(updated_node, graph)
-                        if updated_node_priority < math.inf:
-                            heapq.heappush(
-                                queue, (updated_node_priority, tie_breaker, updated_node)
-                            )
-                            tie_breaker += 1
-
-                # Log str
-                action = f"visited, {len(new_nodes)} new nodes created, {len(nodes_updated)} nodes updated)"
+            # Take nodes from the priority queue until a node eligible for expansion is found
+            found_node = False
+            num_popped = 0
+            while len(queue) > 0 and not found_node:
+                pq_item = queue.pop_item()
+                num_popped += 1
+
+                # Do a few checks
+                assert pq_item.priority < math.inf, "inf priority should not be in the queue"
+                assert math.isclose(
+                    pq_item.priority, self.priority_function(pq_item.item, graph)
+                ), "priority is not up-to-date. This should not happen."
+
+                if self.node_eligible_for_queue(pq_item.item, graph):
+                    found_node = True
+                    node = pq_item.item
+            if not found_node:
+                logger.log(log_level, "No eligible node found. Stopping search.")
+                break
```

Comment:
Could we equivalently use `while ... else` instead of having the `found_node` variable?
```suggestion
            num_popped = 0
            while len(queue) > 0:
                pq_item = queue.pop_item()
                num_popped += 1

                # Do a few checks
                assert pq_item.priority < math.inf, "inf priority should not be in the queue"
                assert math.isclose(
                    pq_item.priority, self.priority_function(pq_item.item, graph)
                ), "priority is not up-to-date. This should not happen."

                if self.node_eligible_for_queue(pq_item.item, graph):
                    node = pq_item.item
                    break
            else:
                logger.log(log_level, "No eligible node found. Stopping search.")
                break
```

Reply:
Yes, that's a good suggestion!